File: backend/app/schemas/extrato_anotado.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, model_validator

StatusRevisao = Literal["pendente", "em_revisao", "revisado", "ignorado"]


class LancamentoExtratoAnotadoResponse(BaseModel):
    id: uuid.UUID
    fechamento_id: uuid.UUID
    arquivo_id: uuid.UUID
    data_lancamento: date
    descricao_banco: str
    razao_social: str | None
    documento: str | None
    valor: Decimal
    tipo_movimento: str
    saldo: Decimal | None
    linha_origem: int | None
    categoria: str | None
    descricao_negocio: str | None
    nf_doc: str | None
    valor_nf_doc: Decimal | None
    observacao: str | None
    categoria_sugerida: str | None
    confianca_sugestao: Decimal | None
    status_revisao: StatusRevisao
    atualizado_por_usuario_id: uuid.UUID | None
    criado_em: datetime
    atualizado_em: datetime

    # Conferência com fluxo de caixa
    previsto_no_fluxo: bool | None = None
    tipo_conferencia_fluxo: str | None = None
    confianca_conferencia: Decimal | None = None
    observacao_sistema: str | None = None
    data_prevista: date | None = None
    valor_previsto: Decimal | None = None
    descricao_prevista: str | None = None

    model_config = {"from_attributes": True}


class AtualizarLancamentoAnotado(BaseModel):
    categoria: str | None = None
    descricao_negocio: str | None = None
    nf_doc: str | None = None
    valor_nf_doc: Decimal | None = None
    observacao: str | None = None
    status_revisao: StatusRevisao | None = None

    @model_validator(mode="after")
    def ao_menos_um_campo(self) -> "AtualizarLancamentoAnotado":
        campos = [self.categoria, self.descricao_negocio, self.nf_doc, self.valor_nf_doc, self.observacao, self.status_revisao]
        if all(c is None for c in campos):
            raise ValueError("Pelo menos um campo deve ser informado.")
        return self

File: frontend/app/types/extratoAnotado.ts
export type StatusRevisao = 'pendente' | 'em_revisao' | 'revisado' | 'ignorado'

export type TipoConferenciaFluxo =
  | 'encontrado'
  | 'data_diferente'
  | 'nao_encontrado'
  | 'valor_diferente'
  | 'possivel_correspondencia'
  | 'pendente_analise'

export interface LancamentoExtratoAnotado {
  id: string
  fechamento_id: string
  arquivo_id: string
  data_lancamento: string
  descricao_banco: string
  razao_social: string | null
  documento: string | null
  valor: number
  tipo_movimento: 'entrada' | 'saida'
  saldo: number | null
  linha_origem: number | null
  categoria: string | null
  descricao_negocio: string | null
  nf_doc: string | null
  valor_nf_doc: number | null
  observacao: string | null
  categoria_sugerida: string | null
  confianca_sugestao: number | null
  status_revisao: StatusRevisao
  atualizado_por_usuario_id: string | null
  criado_em: string
  atualizado_em: string
  // Conferência com fluxo de caixa
  previsto_no_fluxo: boolean | null
  tipo_conferencia_fluxo: TipoConferenciaFluxo | null
  confianca_conferencia: number | null
  observacao_sistema: string | null
  data_prevista: string | null
  valor_previsto: number | null
  descricao_prevista: string | null
}

export interface AtualizarLancamentoAnotado {
  categoria?: string | null
  descricao_negocio?: string | null
  nf_doc?: string | null
  valor_nf_doc?: number | null
  observacao?: string | null
  status_revisao?: StatusRevisao | null
}

export const STATUS_REVISAO_LABELS: Record<StatusRevisao, string> = {
  pendente:   'Pendente',
  em_revisao: 'Em revisão',
  revisado:   'Revisado',
  ignorado:   'Ignorado',
}

// Labels curtos para a tabela
export const CONFERENCIA_LABELS: Record<TipoConferenciaFluxo, string> = {
  encontrado:               'No fluxo',
  data_diferente:           'Data dif.',
  nao_encontrado:           'Não prev.',
  valor_diferente:          'Valor dif.',
  possivel_correspondencia: 'Revisar',
  pendente_analise:         'Pendente',
}

// Labels completos para o modal e exportação
export const CONFERENCIA_LABELS_MODAL: Record<TipoConferenciaFluxo, string> = {
  encontrado:               'Previsto no fluxo',
  data_diferente:           'Data diferente',
  nao_encontrado:           'Não encontrado',
  valor_diferente:          'Valor diferente',
  possivel_correspondencia: 'Revisar',
  pendente_analise:         'Pendente',
}

export const CONFERENCIA_CLASSES: Record<TipoConferenciaFluxo, string> = {
  encontrado:               'bg-green-100 text-green-700',
  data_diferente:           'bg-amber-100 text-amber-700',
  nao_encontrado:           'bg-red-100 text-red-600',
  valor_diferente:          'bg-orange-100 text-orange-700',
  possivel_correspondencia: 'bg-yellow-100 text-yellow-700',
  pendente_analise:         'bg-slate-100 text-slate-500',
}
